### DISCUSSION TITLE
Suggested Faces

### DISCUSSION THREAD

**Discussion by tdoyle3** [RhetTbull/osxphotos#533] [Q&A]
Does anyone know where I can search in the Photos database for which people have face recommendations? I can go to each person in my DB and select the "Confirm Additional Photos" link, but I have 1400 people and this would take forever. What is the query to find those people where potential faces are ready to be reviewed?

**Comment by RhetTbull** [comment]
This might work with osxphotos

`osxphotos query --person _UNKNOWN_ --add-to-album "Faces to be reviewed"`

This query will find all photos that have an unidentified person (e.g. not yet associated with a name) and add them to the album "Faces to be reviewed"

**Comment by tdoyle3** [comment]
Thank you for this answer. While this will give me all photos with an unidentified person, it's not quite what I am searching for. What I'd really like to find is the subset of unidentified persons where photos AI/ML has predicted the face to possibly be of a particular person, and then to get the list of all such persons with potential faces.

**Reply by RhetTbull** [reply]
@tdoyle3 Ah, I understand now. I don't know of way to currently do what you're asking.  I'll take a look around the database when I get a chance and see if I can figure out where Photos stores this information. `ZDEFERREDREBUILDFACE` table might be a good start.

**Comment by tdoyle3** [comment]
Thank you. ZDEFERREDREBUILDFACE is not it though, as I have a copy of my DB where I know I have such tagged potential faces, but this table is empty. Do you know if there is any way to monitor how Photos is interacting with the sqlite DB?

**Reply by RhetTbull** [reply]
I have my own tooling for this to snapshot the database then do a diff. I'll post it here but can't get to it until later tomorrow. Basically, copy the photos.SQLite then make your change in Photos then do a diff with https://www.sqlite.org/sqldiff.html

**Reply by tdoyle3** [reply]
AWESOME!

**Reply by RhetTbull** [reply]
Here's two scripts I use for the development of osxphotos to simplify taking snapshots of the Photos database and doing a diff.

1. Install sqldiff: `brew install sqldiff` (assumes you have [homebrew](https://brew.sh/) installed
2. Unzip the two python scripts from the zip file below and edit each to fix the python path and path to where you want to store the database snapshots.
3. run `plsnap.py`
4. make edits your Photos library
5. run `plsnap.py` again
6. run `pldiff.py` to see a diff of the before/after snapshots

Note: pldiff.py will print out the SQL diff with syntax highlighting but may cut off lines.  You can try it with --raw to disable syntax highlighting and just dump the raw diff text.

[photos_diff.zip](https://github.com/RhetTbull/osxphotos/files/7394257/photos_diff.zip)

**Comment by zephyr325** [comment]
Hey, folks - I've actually got the same situation/question. Did y'all ever get anywhere on this?

**Reply by RhetTbull** [reply]
I've not yet figured this out ( but haven't spent much time looking). There's an association between `ZPERSON`, `ZDETECTEDFACE`, and `ZDETECTEDFACEGROUP` to track these but again, I haven't decoded it.

The following database tables are important:

`ZPERSON` stores information on the person (name, etc).
`ZDETECTEDFACE` stores information the detected face themselves  (eyes open/closed, etc.)
`ZDETECTEDFACEPRINT` stores the "fingerprint" for actual faces
`ZASSET` stores information on each asset (photo or video) in the library
`ZDETECTEDFACEGROUP`: osxphotos doesn't currently use this table but I believe this is the key to the "suggested faces" feature

`ZDETECTEDFACE.ZASSET` is the foreign key to `ZASSET.Z_PK` 
`ZDETECTEDFACE.ZPERSON` is the foriegn key to `ZPERSON.Z_PK`
`ZDETECTEDFACEGROUP.ZASSOCIATEDPERSON` looks like it might be the foreign key to `ZPERSON.Z_PK`

The code osxphotos uses to extract faces is [here](https://github.com/RhetTbull/osxphotos/blob/master/osxphotos/photosdb/_photosdb_process_faceinfo.py) which can give an idea of the SQL used to extract face data (but the code may not make much sense if you don't have a strong understanding of osxphotos internals). 

osxphotos includes built in tools to help with this sort of reverse engineering, namely `osxphotos snap` and `osxphotos diff`.

Here's what I recommend if you want to explore this yourself:

- Open Photos
- Run `osxphotos snap` which creates a snapshot of your Photos database in `/private/tmp/osxphotos_snapshots`
- Select `People`, pick a person, and click "Confirm Additional Photos".
- Confirm one single photo of the person and select `Done`
- Run `osxphotos diff` to see what changed in the database; note `osxphotos diff -r > diff.sql` might be helpful as it prints the "raw" diff whereas the ordinary diff command uses colored text for easier reading but can truncate text -- it's for quick examination whilst `-r` is for deep exploration. 
- Examine the diff output and see if you can figure out what's going on
- Repeat

**Comment by eecue** [comment]
this has a lot of overlap with what i was discussing in the decorator thread… any further movement here?

**Reply by RhetTbull** [reply]
I've not looked into this more since the last post. Doing some sleuthing with the snap and diff commands would be a start but I've had other priorities.
